### PR TITLE
feat: add refresh token interceptor on Suppression API requests

### PR DESCRIPTION
**Pull request by n-r-b**
### JIRA

https://companieshouse.atlassian.net/browse/SR-129

### Change Description

Created refresh access token interceptor for Suppression API requests.

How this works:
- Request interceptor: Sets Axios headers on request.
- Response interceptor: Checks for 401 unauthorised response from API request, and refreshes token. Axios header is updated with new access token. Request is re-tried

### Page Screenshot
N/A

### Work Checklist

- [x] Tests added where applicable
- [x] Test coverage of new code meets target of 80%
- [x] UI changes look good on mobile
- [x] UI changes meet accessibility criteria

---

### Merge Instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/DocumentDetailsController.ts [hunk 1]
```diff
@@ -89,9 +90,10 @@ export class DocumentDetailsController {
     const partialSuppressionData: SuppressionData = { documentDetails } as SuppressionData;
 
     const accessToken: string = SessionService.getAccessToken(req);
+    const refreshToken: string = SessionService.getRefreshToken(req);
 
     try {
-      await this.suppressionService.patch(partialSuppressionData, session.applicationReference, accessToken)
+      await this.suppressionService.patch(partialSuppressionData, session.applicationReference, accessToken, refreshToken)
```
Comment (rhysbarrett) [inline src/controllers/DocumentDetailsController.ts:96]:
```suggestion
      await this.suppressionService.patch(partialSuppressionData, session.applicationReference, accessToken, refreshToken);
```

**Review by akaJSaunders — APPROVED** [review]

File: .env.test [hunk 2]
```diff
@@ -1,6 +1,5 @@
 CDN_HOST=chs-dev.internal:8080
 CHS_URL=http://web.chs-dev.internal:4000
-SUPPRESSIONS_API_URL=http://api.chs-dev.internal
```
Comment (rhysbarrett) [inline .env.test:3]: How come this has been removed?

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.OK, refreshTokenData);
+
+      nock(mockApiHost)
+        .post(mockSuppressionUri, JSON.stringify(mockSuppressionData), {
+            reqheaders: getRequestHeaders(mockNewAccessToken)
+          }
+        ).reply(StatusCodes.CREATED, mockGeneratedReference, {location: '/suppressions/123123'});
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:70]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        .post(mockSuppressionUri, JSON.stringify(mockSuppressionData), {
+            reqheaders: getRequestHeaders(mockAccessToken)
+          }
+        ).reply(StatusCodes.UNAUTHORIZED);
+
+      const refreshTokenMock = nock(mockApiHost)
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.BAD_REQUEST);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:91]: Could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.OK, refreshTokenData);
+
+      nock(mockApiHost)
+        .post(mockSuppressionUri, JSON.stringify(mockSuppressionData), {
+            reqheaders: getRequestHeaders(mockNewAccessToken)
+          }
+        ).reply(StatusCodes.UNAUTHORIZED);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:117]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.OK, refreshTokenData);
+
+      nock(mockApiHost)
+        .get(`${mockSuppressionUri}/${mockGeneratedReference}`, undefined, {
+            reqheaders: getRequestHeaders(mockNewAccessToken)
+          }
+        ).reply(StatusCodes.OK, mockSuppressionData);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:150]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        .get(`${mockSuppressionUri}/${mockGeneratedReference}`, undefined, {
+            reqheaders: getRequestHeaders(mockAccessToken)
+          }
+        ).reply(StatusCodes.UNAUTHORIZED);
+
+      const refreshTokenMock = nock(mockApiHost)
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.BAD_REQUEST);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:171]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        new RefreshTokenService(mockApiHost + mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret));
+
+      await suppressionService.get(mockGeneratedReference, mockAccessToken, mockRefreshToken).then((response: SuppressionData) => {
+        expect(refreshTokenMock.isDone()).toBeTruthy();
+        expect(response).toEqual(mockSuppressionData)
+      });
+    });
+
+    it('should return status 400 when invalid token refresh fails with 400', async () => {
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:159]: not sure I like the name of this test, maybe be more descriptive on what 400 failure means

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+      const suppressionService = new SuppressionService(mockApiHost,
+        new RefreshTokenService(mockApiHost + mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret));
+
+      await suppressionService.patch(mockPartialData, mockGeneratedReference, mockAccessToken, mockRefreshToken).then(response => {
+        expect(refreshTokenMock.isDone()).toBeTruthy();
+        expect(response).toBeUndefined();
+      });
+    });
+
+    it('should return status 400 when expired token refresh fails with 400', async () => {
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:240]: not sure I like the name of this test, maybe be more descriptive on what 400 failure means

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.OK, refreshTokenData);
+
+      nock(mockApiHost)
+        .patch(mockSuppressionUri + `/${mockGeneratedReference}`, undefined, {
+            reqheaders: getRequestHeaders(mockNewAccessToken)
+          }
+        ).reply(StatusCodes.NO_CONTENT);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:231]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        .patch(`${mockSuppressionUri}/${mockGeneratedReference}`, undefined, {
+            reqheaders: getRequestHeaders(mockAccessToken)
+          }
+        ).reply(StatusCodes.UNAUTHORIZED);
+
+      const refreshTokenMock = nock(mockApiHost)
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.BAD_REQUEST);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:252]: could we add the type here

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        .post(`${mockRefreshServiceUri}${mockRefreshParams}`)
+        .reply(StatusCodes.OK, refreshTokenData);
+
+      nock(mockApiHost)
+        .patch(`${mockSuppressionUri}/${mockGeneratedReference}`, undefined, {
+            reqheaders: getRequestHeaders(mockNewAccessToken)
+          }
+        ).reply(StatusCodes.UNAUTHORIZED);
+
+      const suppressionService = new SuppressionService(mockApiHost,
```
Comment (akaJSaunders) [inline test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts:279]: could we add the type here

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+  const refreshTokenData: RefreshTokenData = {
+    'expires_in': 3600,
+    'token_type': 'Bearer',
+    'access_token': `${mockNewAccessToken}`
+  };
+
+  describe('Refreshing token', () => {
+    it('should throw an error when Access token not defined', async () => {
+      const refreshTokenService = new RefreshTokenService(mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret);
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:30]: could we add the type here

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+      for (const data of [undefined, null]) {
+        await refreshTokenService.refresh(data as any, mockRefreshToken).catch((err) => {
+          expect(err).toEqual(Error('Access token is missing'))
+        });
+      }
+    });
+
+    it('should throw an error when refresh token not defined', async () => {
+      const refreshTokenService = new RefreshTokenService(mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret);
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:40]: could we add the type here

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+      for (const data of [undefined, null]) {
+        await refreshTokenService.refresh(mockAccessToken, data as any).catch((err) => {
+          expect(err).toEqual(Error('Refresh token is missing'))
+        });
+      }
+    });
+
+    it('should refresh access token', async () => {
+
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:50]:
```suggestion
```

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+      mockedAxios.post.mockImplementationOnce(() => Promise.resolve({
+        status: StatusCodes.OK,
+        data: refreshTokenData
+      }));
+
+      const refreshTokenService = new RefreshTokenService(mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret);
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:56]: could we add the type here

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+      await refreshTokenService.refresh(mockAccessToken, mockRefreshToken).then((response: string) => {
+        expect(response).toEqual(mockNewAccessToken)
+      });
+
+    });
+
+    it('should throw error when response is empty', async () => {
+
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:65]:
```suggestion
```

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+      mockedAxios.post.mockReturnValue(Promise.resolve({
+        status: StatusCodes.OK
+      }));
+
+      const refreshTokenService = new RefreshTokenService(mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret);
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:70]: could we add the type here

File: test/services/refresh-token/RefreshTokenService.test.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      const errorResponse = {
+        message: 'Request failed with status code 400',
+        response: {
+          status: StatusCodes.BAD_REQUEST
+        }
+      };
+
+      mockedAxios.post.mockReturnValue(Promise.reject(errorResponse));
+
+      const refreshTokenService = new RefreshTokenService(mockRefreshServiceUri, mockRefreshClientId, mockRefreshClientSecret);
```
Comment (akaJSaunders) [inline test/services/refresh-token/RefreshTokenService.test.ts:89]: could we add the type here

File: test/services/session/SessionService.test.ts [hunk 20]
```diff
@@ -191,6 +191,28 @@ describe('SessionService', () => {
     const result = SessionService.getUserEmail(mockRequest);
     expect(result).toEqual(testEmail);
     expect(mockGetSignInInfo).toHaveBeenCalledWith(SessionKey.SignInInfo);
-  })
+  });
+
+  it('should retrieve the refresh token from the session', () => {
+
+    const testToken = 'test-token';
```
Comment (akaJSaunders) [inline test/services/session/SessionService.test.ts:198]: could we add the type here

File: test/services/session/SessionService.test.ts [hunk 21]
```diff
@@ -169,9 +169,9 @@ describe('SessionService', () => {
     const result = SessionService.getAccessToken(mockRequest);
     expect(result).toEqual(testToken);
     expect(mockGetSignInInfo).toHaveBeenCalledWith(SessionKey.SignInInfo);
-  })
+  });
 
-  it('should retrieve the access token from the session', () => {
+  it('should retrieve the email address from the session', () => {
 
     const testEmail = 'test@example.com';
```
Comment (akaJSaunders) [inline test/services/session/SessionService.test.ts:176]: could we add the type here

File: test/services/session/SessionService.test.ts [hunk 22]
```diff
@@ -191,6 +191,28 @@ describe('SessionService', () => {
...
+          refresh_token: testToken
+        },
+        user_profile: {
+          email: 'test@example.com'
+        }
+      }
+    });
+    mockRequest.session!.get = mockGetSignInInfo;
+
+    const result = SessionService.getRefreshToken(mockRequest);
```
Comment (akaJSaunders) [inline test/services/session/SessionService.test.ts:213]: could we add the type here

**Review by akaJSaunders — COMMENTED** [review]
Good job!! 🥇  a few comments regarding little details about the tests

File: .env.test (hunk 2 above)
Comment (n-r-b) [reply .env.test:3]: it's a duplicate

File: test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts (hunk 8 above)
Comment (n-r-b) [reply test/services/refresh-token/AppliedRefreshTokenInterceptor.test.ts]: i've changed the name, let me know if it still doesn't make sense

**Review by akaJSaunders — APPROVED** [review]
💯

**Pull request merged**